app/lib/logout_token.rb

```diff
@@ -0,0 +1,124 @@
+require "attr_required"
+require "attr_optional"
+
+class LogoutToken
+  include AttrOptional
+  include AttrRequired
+  class InvalidToken < RuntimeError; end
+  class ExpiredToken < InvalidToken; end
+  class InvalidIssuer < InvalidToken; end
+  class InvalidAudience < InvalidToken; end
+  class InvalidIssuedAt < InvalidToken; end
+  class InvalidEvent < InvalidToken; end
+  class InvalidBackchanelLogoutEvent < InvalidToken; end
+  class InvalidIdentifiers < InvalidToken; end
+  class TokenRecentlyUsed < InvalidToken; end
+  class ProhibitedNonse < InvalidToken; end
+
+  BACK_CHANNEL_EVENT_NAME = "http://schemas.openid.net/event/backchannel-logout".freeze
+
+  attr_required :iss, :aud, :iat, :jti, :events
+  attr_optional :sub, :sid, :auth_time
+  attr_accessor :access_token, :code, :state
+  alias_method :subject, :sub
+  alias_method :subject=, :sub=
+
+  NON_STRING_ATTRIBUTES = %i[aud exp iat auth_time sub_jwk events].freeze
+
+  def initialize(attributes = {})
+    attributes_to_methods(attributes)
+    apply_attribute_values_types(all_attributes)
+    validate_prohibited_attribtues(attributes)
```

Comment:
Interesting that we're validating the `nonse` here but all the other attributes only get checked when we call `verify!` What's the reason for that?

Reply:
So I'll just play back out chat here for posteirty. It's a good question.
There's an OK reason for this and a bit of a naff reason.

## OK reason: Validating attributes vs Validating values

The OK reason is that there is a split in validation here:
I based the `logout_token.rb` on the [id_token.rb](https://github.com/nov/openid_connect/blob/master/lib/openid_connect/response_object/id_token.rb) as it is in[ our dependency](https://github.com/alphagov/account-api/blob/main/Gemfile#L20).

The author there created their [own gem](https://github.com/nov/attr_required) and used it in their [openid_connect id_token](https://github.com/nov/openid_connect/blob/master/lib/openid_connect/response_object/id_token.rb#L10-L11). The theory is, don't initalise the object without throwing an error if the core named parameters are not present.

I've re-used that here.

So then why validate prohibited attribtues here? The[ back_channel_logout spec](https://openid.net/specs/openid-connect-backchannel-1_0.html#Validation) has an odd bit:

> Verify that the Logout Token does not contain a nonce Claim. 

So we've got something `attr_required` doesn't cater for here as a prohibited attribute, so I rolled my own.

In general I'm seeing the difference as:
- In initalize check the attributes are present / as expected enough to be happy to continue
- verify! check the values and confirm if it's valid

I'm not entirely sure I like this pattern, GOV.UK folks always instilled in me that we should keep very little going on in initalize... Here we've got quite a lot.

## Rubbish Reason: So why keep it this way?

I kept it this way becuase I wanted to keep things looking roughly like the openid_connect gem. This is a rubbish reason because I'm favouring the style / format of a dependency instead of how we've written things elsewhere in this codebase... which feels a bit... eww.

But it made sense when writing the logout token as [the spec](https://openid.net/specs/openid-connect-backchannel-1_0.html#Validation) mentions a bunch of things that should be verified in the same way we do with an ID token. And I'm still hoping the discussion will continue on an issue I raised in [the dependency](https://github.com/nov/openid_connect/issues/67). In an ideal world, I could contribute some portion of this code back as a PR, and then we could outsource this token code to a module that comes with the ID token checks.

This doesn't seem like the most compelling of arguments, but it's where I ended up, and I don't think the cost of a somewhat busy initaliser is too bad, and I think the code is legible enough to merge.

Open to 🧠 if you disagree!

Reply:
No, that sounds good to me 👍🏻